### PR TITLE
Add result property of type unknown to the VFile interface definition…

### PR DESCRIPTION
**Pull request by gurkerl83**
…. Add property description, adjust readme.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -53,7 +53,7 @@
     "remark-preset-wooorm": "^6.0.0",
     "tape": "^4.0.0",
     "tinyify": "^2.0.0",
-    "xo": "^0.27.0"
+    "xo": "0.27.0"
```
Comment (wooorm) [inline package.json:56]:
```suggestion
    "xo": "^0.27.0"
```

Could you revert this change?

Comment (wooorm) [reply package.json]: I pushed updates to the dev dependencies so this should no longer be an issue

Comment (gurkerl83) [reply package.json]: This change is fine with me. It seems the change you made solved this issue. Still, I see that one of the last test pipelines was not executed successfully. https://travis-ci.org/github/vfile/vfile/builds/671479197?utm_source=github_status&utm_medium=notification

Comment (wooorm) [reply package.json]: CI failures seem to be unrelated: Both GH and Travis fail on installing Node for some reason.

Comment (gurkerl83) [reply package.json]: About the other question of not defining the "result" key at all. The description of the data key is misleading, in my opinion. Instead of writing data?: unknown and setting the type to "unknown," you could have set the type to "any." Hiding information behind an even more complex data structure is then possible. The only option you have in typescript to not define a property (key in an object) at all is to add "[key: string]: any" to an interface VFile the same as in VFileOptions. But there are problems with that approach, some of with Intelli sense of IDEs. Also, typescript set to a stricter mode wouldn't probably allow this kind of declaration.

Comment (wooorm) [reply package.json]: Yes, the description of allowing anything directly on `file` should indeed not be in `file.data`. I think a new documentation section defining common stuff set on `file`, like how it’s done in `vfile-message` in its [Custom properties](https://github.com/vfile/vfile-message#custom-properties) section, makes sense.
I can take that on and do docs later!

I’m indeed thinking of `[key: string]: any` on `file`. This is already what passing a random field in `VFileOptions` will result in.

**Comment by Rokt33r** [comment]
@wooorm 
> Is it somehow possible with typescript to not define .result, but instead allow any value to be defined

It's possible. 
We did same thing for unist's `Node`. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/unist/index.d.ts#L31

```ts
interface VFile {
  ... // Other props
  [key: string]: unknown
}
```

**Comment by wooorm** [comment]
@gurkerl83 I think I’d prefer that (@Rokt33r above) on the types’ side for this PR, and then I can solve the docs by mentioning `.result` (and other fields?)

**Comment by gurkerl83** [comment]
This is fine with me. Using this approach, you will lose the Inteli-sense feature of IDEs. The result property will not reveal itself in the options list of a VFile object. I was thinking about the general setup and maybe a sort of duplication of fields defined in the VFile interface. Looking at the unified project and the other ongoing discussion about contents -> value, it is my impression that contents and result fields both serve as output returned from a processor function. The idea is to allow an external library to define T, what the expected return type will be. Of course, this will be some sort of breaking change, but something which is a bit more deterministic about what the input, output, and options are would be awesome - something like that.

```
interface VFile<T> {
  contents: VFileContents;
  result: {
    contents: VFileContents;
    value: T;
  };
}
```

**Comment by Rokt33r** [comment]
@gurkerl83 I feel a bit lost. Which packages are using `.result` now? I think we should define something like `VFileWithResult`, which having a definition for `.result`, in those packages rather than defining it in vfile.

**Comment by gurkerl83** [comment]
It was not my intention to say that other libraries do anything with result, because it was introduced in unified only a week ago. My current state of knowledge about the inner workings of unifieds middleware is very shallow. I have to take a closer look at pipelineStringify, but more explicitly at the processors stringify function. The use of the generic type T mentioned in the last comment should only indicated that a Compiler defined from the outside and attached to unified processor knows its expected return type and maybe the this corresponds type of the result field which maybe can be inferred. If this comment does make any sense because of my previous mentioned lack of understanding please ignore it and go forward with the proposed change of using typescripts wildcard approach.

**Comment by Rokt33r** [comment]
I see. Now I understand. In that case, I think we should just go for the wildcard approach and fix typings of unified so adopters can customize vfile typing from `.process` or `.procesSync`. I think providing generic to those methods should be easier to implement and to use.

```ts
interface CustomVFile extends VFile {
  result: ...
}

const processor = unified().use(...).use(...)

function process(input: VFileCompatible) {
  return processor.process<CustomVFile>(input)
}
```

**Comment by wooorm** [comment]
Per @Rokt33r’s last comment, does this mean this PR can be closed? Should something change?

**Comment by chrisrzhou** [comment]
Any updates on this? I don't have too much context on how this would impact the library itself (doesn't seem like a backwards incompatible change) so I'm assuming it's just confirmation on `unknown` being the correct type definition for `vfile.result`?

Shipping this would be great to resolve a couple of type issues coming from accessing `result` on `vfile`.

**Comment by Rokt33r** [comment]
@wooorm @ChristianMurphy 
I think we can close this pr and introduce the wild card approach(https://github.com/vfile/vfile/pull/53#issuecomment-611879864).

Although we cannot catch any mistakes like the below, the approach makes using unified packages with typescript easier for everyone. People like me, who use typescript strictly, can introduce extra types and tests outside of this package very easily.

```ts
const file = unified().use(...).processSync(...) // returns `VFile`
console.log(file.result) // No error because any extra props of `VFile` is unknown
console.log(file.resu) // No error because of the same reason
```

**Comment by Rokt33r** [comment]
I've submitted https://github.com/vfile/vfile/pull/55

**Comment by wooorm** [comment]
Released in 4.2.0!
